src/treelearner/feature_histogram.hpp

```diff
@@ -457,38 +506,49 @@ class FeatureHistogram {
     }
   }
 
- private:
-  static double GetSplitGains(double sum_left_gradients, double sum_left_hessians,
-                              double sum_right_gradients, double sum_right_hessians,
-                              double l1, double l2, double max_delta_step,
-                              double min_constraint, double max_constraint, int8_t monotone_constraint) {
-    double left_output = CalculateSplittedLeafOutput(sum_left_gradients, sum_left_hessians, l1, l2, max_delta_step, min_constraint, max_constraint);
-    double right_output = CalculateSplittedLeafOutput(sum_right_gradients, sum_right_hessians, l1, l2, max_delta_step, min_constraint, max_constraint);
-    if (((monotone_constraint > 0) && (left_output > right_output)) ||
-      ((monotone_constraint < 0) && (left_output < right_output))) {
-      return 0;
-    }
-    return GetLeafSplitGainGivenOutput(sum_left_gradients, sum_left_hessians, l1, l2, left_output)
-      + GetLeafSplitGainGivenOutput(sum_right_gradients, sum_right_hessians, l1, l2, right_output);
-  }
-
   /*!
   * \brief Calculate the output of a leaf based on regularized sum_gradients and sum_hessians
   * \param sum_gradients
   * \param sum_hessians
   * \return leaf output
   */
+  template <class T>
   static double CalculateSplittedLeafOutput(double sum_gradients, double sum_hessians, double l1, double l2, double max_delta_step,
-                                            double min_constraint, double max_constraint) {
+                                            const T *constraint) {
     double ret = CalculateSplittedLeafOutput(sum_gradients, sum_hessians, l1, l2, max_delta_step);
-    if (ret < min_constraint) {
-      ret = min_constraint;
-    } else if (ret > max_constraint) {
-      ret = max_constraint;
+    if (ret < constraint->GetCurrentMinConstraint()) {
+      ret = constraint->GetCurrentMinConstraint();
+    } else if (ret > constraint->GetCurrentMaxConstraint()) {
+      ret = constraint->GetCurrentMaxConstraint();
     }
     return ret;
   }
 
+ private:
+  static double GetSplitGains(double sum_left_gradients, double sum_left_hessians,
+                              double sum_right_gradients, double sum_right_hessians,
+                              double l1, double l2, double max_delta_step,
+                              const SplittingConstraints *constraints, int8_t monotone_constraint) {
+    const SplittingConstraint *left;
+    const SplittingConstraint *right;
+    if (constraints != nullptr) {
+      left = &(constraints->left);
+      right = &(constraints->right);
+    }
+    else {
+      left =  new SplittingConstraint();
+      right =  new SplittingConstraint();
```

Comment:
maybe initialize left and right to nullptr? and handle the nullptr in CalculateSplittedLeafOutput function?